`passwordChecklist.js` Changes

aaronskiba opened the pull request:
### Changes proposed in this PR:
[Add const to passwordRequirements](https://github.com/portagenetwork/roadmap/commit/90181089eb0114df24dd3231e724bb946699d301)

[Wrap passwordChecklist.js in function / enable early return](https://github.com/portagenetwork/roadmap/commit/e9c9e287a6f1c071498a93c32b93629b789a1559)
- This change adds the early return of `passwordChecklist.js` when a page doesn't include any relevant elements (i.e. `if (passwordInputs.length === 0 || !passwordRequirements) return;`).
- To enable this early return, we are now assigning the prior code to the `passwordChecklist` function, and then calling it immediately.

[Update passwordChecklist.js to reduce DOM queries](https://github.com/portagenetwork/roadmap/commit/8f330c18f2b32fc5e177e750f7765831e1536122)
- Add `ruleIconsMap` to map password rules to their corresponding `<i>` icons.
- Eliminates repeated DOM queries previously being performed in `updateRequirements()`

[Simplify passwordChecklist.js to single password input](https://github.com/portagenetwork/roadmap/commit/c76393901188fee0ab7d6787e713bd557b4399a6)
- Replaced the `passwordsInputs` array with a single `passwordInput` variable. Pages never have more than one password field that uses `passwordField.js`, and this change should simplify the code a little bit.

[Rename variables in passwordChecklist.js](https://github.com/portagenetwork/roadmap/commit/bbdde6cf034a6d3cef44c85e6a806544b3e00e62)
- Renamed `passwordRequirements` to `passwordChecklist` to reflect elements actual name
- Renamed `passwordChecklist` function name to `initPasswordChecklist` to better distinguish it from the newly added `passwordChecklist` variable
- Renamed `updateRequirements` to `updateChecklistIcons` to clarify behaviour

[Remove redundant explicit password_field IDs](https://github.com/portagenetwork/roadmap/pull/1162/commits/e465fac00396c768de6fa278c1067e8686c5eac3)
- Rails automatically assigns `id="user_password"` to password fields, so
explicit `id: 'user_password'` attributes were unnecessary.
- Additionally, the manually defined `id: 'user_password'` in `app/views/shared/_create_account_form.html.erb` had no effect. Rather, `id="new_user_password"` was being generated due to `namespace: 'new'` being specified in the form.

aaronskiba committed:
Add `const` to `passwordRequirements`

aaronskiba committed:
Wrap passwordChecklist.js in function / enable early return

- This change adds the early return of `passwordChecklist.js` when a page doesn't include any relevant elements (i.e. `if (passwordInputs.length === 0 || !passwordRequirements) return;`).
- To enable this early return, we are now assigning the prior code to the `passwordChecklist` function, and then calling it immediately.

aaronskiba committed:
Update passwordChecklist.js to reduce DOM queries

Add `ruleIconsMap` to map password rules to their corresponding `<i>` icons.
- Eliminates repeated DOM queries previously being performed in `updateRequirements()`

aaronskiba committed:
Simplify passwordChecklist.js to single password input

Replaced the `passwordsInputs` array with a single `passwordInput` variable. Pages never have more than one password field that uses `passwordField.js`, and this change should simplify the code a little bit.

aaronskiba committed:
Rename variables in `passwordChecklist.js`

- Renamed `passwordRequirements` to `passwordChecklist` to reflect elements actual name
- Renamed `passwordChecklist` function name to `initPasswordChecklist` to better distinguish it from the newly added `passwordChecklist` variable
- Renamed `updateRequirements` to `updateChecklistIcons` to clarify behaviour

momo3404 commented on app/javascript/src/shared/passwordChecklist.js:
```diff
@@ -1,39 +1,50 @@
+  );
+
+  const updateChecklistIcons = (pw) => {
+    for (const [ruleName, ruleValidator] of Object.entries(rules)) {
```
I think a quick comment explaining how this works in practice could be useful, the multiple rule variable names can be a bit overwhelming.

Something like `// for (const [lowercase, (pw) => /[a-z]/.test(pw)] in rules ... icon = lowercase icon`

momo3404 commented on app/javascript/src/shared/passwordChecklist.js:
```diff
@@ -1,39 +1,50 @@
-  }
-};
+  const rules = {
+    length: (pw) => pw.length >= 8,
```
A comment here explaining `.test` could be helpful, as in mentioning that it returns a boolean if `pw` matched the regexp. You could also paste this link if you want (it's what I used when writing this initially) https://www.w3schools.com/jsref/jsref_regexp_test.asp

momo3404 reviewed (approved):
Some great javascript optimization!

aaronskiba committed:
Add comments to `passwordChecklist.js`

aaronskiba committed:
Remove redundant explicit `password_field` IDs

- Rails automatically assigns `id="user_password"` to password fields, so
explicit `id: 'user_password'` attributes were unnecessary.

- Additionally, the manually defined `id: 'user_password'` in `app/views/shared/_create_account_form.html.erb` had no effect. Rather, `id="new_user_password"` was being generated due to `namespace: 'new'` being specified in the form.
